---
title: "Starting with data"
author: "Data Carpentry contributors"
minutes: 25
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(error=TRUE)
```

```{r, echo=FALSE, purl=FALSE, message = FALSE}
source("setup.R")
```

------------

> ### Learning Objectives
>
> * Describe what an R data frame is.
> * Load external data from a .csv file into R.
> * Summarize the contents of a data frame in R.
> * Manipulate categorical data in R.

------------

# Looking at Metadata

```{r, echo=FALSE, purl=TRUE}
# Looking at metadata
```

We will be looking at the metadata of a gene knockout study. The aim of this study was to determine whether knocking out the protein 'chemerin' affects gut microbial composition. In total, 116 mouse samples were acquired from two different facilities. Metadata associated with each sample is indicated in the mapping file (mouse_mapfile.txt), which we will download using RStudio. In this mapping file the genotypes of interest can be seen: wildtype (WT), chemerin knockout (chemerin_KO), chemerin receptor knockout (CMKLR1_KO) and a heterozygote for the receptor knockout (HET). Also of importance are the two source facilities: "BZ" and "CJS". It is generally a good idea to include as much metadata as possible, since this data can easily be explored later on.

The metadata mapping file of this dataset is stored as a comma separated value (CSV) file.
Each row holds information for a single mouse, and the columns represent:

| Column                | Description                                           |
|-----------------------|-------------------------------------------------------|
| SampleID              | Unique id for each sample                             |
| BarcodeSequence       | Barcode sequence used for each sample (NA)            |
| LinkerPrimerSequence  | Linker/primer sequence used to amplify the sample (NA)|
| FileInput             | Name of .fastq input file                             |
| Source                | BZ or CJS                                             |
| Mouse                 | Unique number for each mouse                          |
| Cage                  | Cage number (NA)                                      |
| Genotype              | WT, chemerin_KO, CMKLR1_HE, or CMKLR1_KO              |
| SamplingWeek          | Week in which mouse was sampled                       |
| Description           | Unique id for each mouse                              |

We are going to use the R function `download.file()` to download the CSV file
that contains the metadata, and we will use `read.csv()` to
load into memory the content of the CSV file as an object of class `data.frame`.

To download the metadata into the `data/` subdirectory, run the following:


```r
download.file("https://raw.githubusercontent.com/BinxiePeterson/introduction-to-R-for-16S/gh-pages/mouse_mapfile.txt",
              "data/metadata.csv")
```

You are now ready to load the metadata:


```r
metadata <- read.csv('data/metadata.csv', header = TRUE, sep = "\t")
```

This statement doesn't produce any output because, as you might recall,
assignments don't display anything. If we want to check that our data has been
loaded, we can print the variable's value: `metadata`.

This means the data loaded properly. Let's check the top (the first 6 lines) of this data frame using the function `head()`:

```{r, results='show', purl=FALSE}
head(metadata)
```

> ### Note
>
> `read.csv` assumes that fields are delineated by commas, however, in several
> countries, the comma is used as a decimal separator and the semicolon (;) is
> used as a field delineator. If you want to read in this type of files in R,
> you can use the `read.csv2` function. It behaves exactly like `read.csv` but
> uses different parameters for the decimal and the field separators. If you are
> working with another format, they can be both specified by the user. Check out
> the help for `read.csv()` to learn more.


## What are data frames?

Data frames are the _de facto_ data structure for most tabular data, and what we
use for statistics and plotting.

A data frame can be created by hand, but most commonly they are generated by the
functions `read.csv()` or `read.table()`; in other words, when importing
spreadsheets from your hard drive (or the web).

A data frame is the representation of data in the format of a table where the
columns are vectors that all have the same length. Because the column are
vectors, they all contain the same type of data (e.g., characters, integers,
factors). For example, here is a figure depicting a data frame comprising of a
numeric, a character and a logical vector.

![](./img/data-frame.svg)


We can see this when inspecting the <b>str</b>ucture of a data frame
with the function `str()`:

```{r, purl=FALSE}
str(metadata)
```

## Inspecting `data.frame` Objects

We already saw how the functions `head()` and `str()` can be useful to check the
content and the structure of a data frame. Here is a non-exhaustive list of
functions to get a sense of the content/structure of the data. Let's try them out!

* Size:
    * `dim(metadata)` - returns a vector with the number of rows in the first element,
          and the number of columns as the second element (the **dim**ensions of
          the object)
    * `nrow(metadata)` - returns the number of rows
    * `ncol(metadata)` - returns the number of columns

* Content:
    * `head(metadata)` - shows the first 6 rows
    * `tail(metadata)` - shows the last 6 rows

* Names:
    * `names(metadata)` - returns the column names (synonym of `colnames()` for `data.frame`
	   objects)
    * `rownames(metadata)` - returns the row names

* Summary:
    * `str(metadata)` - structure of the object and information about the class, length and
	   content of  each column
    * `summary(metadata)` - summary statistics for each column

Note: most of these functions are "generic", they can be used on other types of
objects besides `data.frame`.


> ### Challenge
>
> Based on the output of `str(metadata)`, can you answer the following questions?
>
> * What is the class of the object `metadata`?
> * How many rows and how many columns are in this object?

```{r, echo=FALSE, purl=TRUE}

## Challenge
## Based on the output of `str(metadata)`, can you answer the following questions?
## * What is the class of the object `metadata`?
## * How many rows and how many columns are in this object?
```

<!---
```{r, echo=FALSE, purl=FALSE}
## Answers
##
## * class: data frame
## * how many rows: 116,  how many columns: 10
```
--->


## Indexing and subsetting data frames

```{r, echo=FALSE, purl=TRUE}

## Indexing and subsetting data frames
```


Our data frame has rows and columns (it has 2 dimensions), if we want to
extract some specific data from it, we need to specify the "coordinates" we
want from it. Row numbers come first, followed by column numbers. However, note
that different ways of specifying these coordinates lead to results with
different classes.

```{r, purl=FALSE}
metadata[1, 1]   # first element in the first column of the data frame (as a vector)
metadata[1, 6]   # first element in the 6th column (as a vector)
metadata[, 1]    # first column in the data frame (as a vector)
metadata[1]      # first column in the data frame (as a data.frame)
metadata[1:3, 7] # first three elements in the 7th column (as a vector)
metadata[3, ]    # the 3rd element for all columns (as a data.frame)
head_metadata <- metadata[1:6, ] # equivalent to head_metadata <- head(metadata)
```

`:` is a special function that creates numeric vectors of integers in increasing
or decreasing order, test `1:10` and `10:1` for instance.

You can also exclude certain parts of a data frame using the "`-`" sign:

```{r, purl=FALSE}
metadata[,-1]          # The whole data frame, except the first column
metadata[-c(7:116),]   # Equivalent to head(metadata)
```

As well as using numeric values to subset a `data.frame` (or `matrix`), columns
can be called by name, using one of the four following notations:

```{r, eval = FALSE, purl=FALSE}
metadata["SampleID"]       # Result is a data.frame
metadata[, "SampleID"]     # Result is a vector
metadata[["SampleID"]]     # Result is a vector
metadata$SampleID          # Result is a vector
```

For our purposes, the last three notations are equivalent. RStudio knows about
the columns in your data frame, so you can take advantage of the autocompletion
feature to get the full and correct column name.

> ### Challenge
>
> 1. Create a `data.frame` (`metadata_80`) containing only the observations from
>    row 80 of the `metadata` dataset.
>
> 2. Notice how `nrow()` gave you the number of rows in a `data.frame`?
>
>      * Use that number to pull out just that last row in the data frame.
>      * Compare that with what you see as the last row using `tail()` to make
>        sure it's meeting expectations.
>      * Pull out that last row using `nrow()` instead of the row number.
>      * Create a new data frame object (`metadata_last`) from that last row.
>
> 3. Use `nrow()` to extract the row that is in the middle of the data
>    frame. Store the content of this row in an object named `metadata_middle`.
>
> 4. Combine `nrow()` with the `-` notation above to reproduce the behavior of
>    `head(metadata)` keeping just the first through 6th rows of the metadata.


```{r, echo=FALSE, purl=TRUE}
### Challenges:
###
### 1. Create a `data.frame` (`metadata_80`) containing only the
###    observations from row 80 of the `metadata`.
###
### 2. Notice how `nrow()` gave you the number of rows in a `data.frame`?
###
###      * Use that number to pull out just that last row in the data frame
###      * Compare that with what you see as the last row using `tail()` to make
###        sure it's meeting expectations.
###      * Pull out that last row using `nrow()` instead of the row number
###      * Create a new data frame object (`metadata_last`) from that last row
###
### 3. Use `nrow()` to extract the row that is in the middle of the
###    data frame. Store the content of this row in an object named
###    `metadata_middle`.
###
### 4. Combine `nrow()` with the `-` notation above to reproduce the behavior of
###    `head(metadata)` keeping just the first through 6th rows of the metadata.

```

<!---
```{r, purl=FALSE}
## Answers
metadata_80 <- metadata[80, ]
metadata_last <- metadata[nrow(metadata), ]
metadata_middle <- metadata[nrow(metadata)/2, ]
metadata_head <- metadata[-c(7:nrow(metadata)),]
```
--->


## Factors

```{r, echo=FALSE, purl=TRUE}
### Factors
```

When we did `str(metadata)` we saw that some of the columns consist of
integers, however, multiple columns, such as `SampleID`, `FileInput`, `Source`, `Genotype`, and `Description` are of a special class called a `factor`. Factors are very useful and are actually something that make R particularly well suited to working with data, so we're
going to spend a little time introducing them.

Factors are used to represent categorical data. Factors can be ordered or
unordered, and understanding them is necessary for statistical analysis and for
plotting.

Factors are stored as integers, and have labels (text) associated with these
unique integers. While factors look (and often behave) like character vectors,
they are actually integers under the hood, and you need to be careful when
treating them like strings.

In the data frame we just imported, let's do 
```{r, purl=TRUE}
str(metadata)
```

We can see the names of the multiple columns. And, we see that 
some say things like `Factor w/ 116 levels`

When we read in a file, any column that contains text is automatically
assumed to be a factor. Once created, factors can only contain a pre-defined set values, known as
*levels*. By default, R always sorts *levels* in alphabetical order. 

For instance, we see that `Genotype` is a Factor w/ 4 levels, `WT`, `chemerin_KO`, `CMKLR1_HE`, and `CMKLR1_KO`.

You can check this by using the function `levels()`, and check the
number of levels using `nlevels()`:


```{r, purl=FALSE, eval=FALSE}
levels(metadata$Genotype)
nlevels(metadata$Genotype)
```

For
instance, if you have a factor with 2 levels:

```{r, purl=TRUE}
sex <- factor(c("male", "female", "female", "male"))
```

R assigned `1` to the level `"female"` and `2` to the level `"mail"` (because
`f` comes before `m`, even though the first element in this vector is
`"male"`). You can check this by using the function `levels()`, and check the
number of levels using `nlevels()`:

```{r, purl=FALSE}
levels(sex)
nlevels(sex)
```

Sometimes, the order of the factors does not matter, other times you might want
to specify the order because it is meaningful (e.g., "low", "medium", "high"),
it improves your visualization, or it is required by a particular type of
analysis. Additionally, specifying the
order of the levels allows to compare levels:

```{r, purl=FALSE, error=TRUE, eval=FALSE}
expression <- factor(c("low", "high", "medium", "high", "low", "medium", "high"))
levels(expression)
expression <- factor(expression, levels=c("low", "medium", "high"))
levels(expression)
min(expression) ## doesn't work
expression <- factor(expression, levels=c("low", "medium", "high"), ordered=TRUE)
levels(expression)
min(expression) ## works!
```

In R's memory, these factors are represented by integers (1, 2, 3), but are more
informative than integers because factors are self describing: `"male"`,
`"female"` is more descriptive than `1`, `2`. Which one is "female"?  You wouldn't
be able to tell just from the integer data. Factors, on the other hand, have
this information built in. It is particularly helpful when there are many levels
(like the SampleID in our example dataset).

### Converting factors

If you need to convert a factor to a character vector, you use
`as.character(x)`.

Converting a factor to a numeric vector is however a little trickier, and you
have to go via a character vector. Compare:

```{r, purl=FALSE}
as.character(metadata$Source)
```

Notice that in the `levels()` approach, three important steps occur:

* We obtain all the factor levels using `levels(f)`
* We convert these levels to numeric values using `as.numeric(levels(f))`
* We then access these numeric values using the underlying integers of the
  vector `f` inside the square brackets


### Basic plotting

When your data is stored as a factor, you can use the `plot()` function to get a
quick glance at the number of observations represented by each factor
level. Let's look at the number of mice per treatment group:

```{r, purl=TRUE}
## bar plot of the number of male and female participants in the study:
plot(metadata$Genotype)
```


### Challenge

The function `table()` tabulates observations and can be used to create
bar plots quickly. For instance:

```{r wrong-order, results='show', purl=TRUE}
## Question: How can you recreate this plot but by having "control"
## being listed last instead of first?
exprmt <- factor(c("treat1", "treat2", "treat1", "treat3", "treat1", "control",
                   "treat1", "treat2", "treat3"))
table(exprmt)
barplot(table(exprmt))
```

<!---
```{r correct-order, purl=FALSE}
exprmt <- factor(exprmt, levels=c("treat1", "treat2", "treat3", "control"))
barplot(table(exprmt))
```
--->

### Using `stringsAsFactors=FALSE`

By default, when building or importing a data frame, the columns that contain
characters (i.e., text) are coerced (=converted) into the `factor` data
type. Depending on what you want to do with the data, you may want to keep these
columns as `character`. To do so, `read.csv()` and `read.table()` have an
argument called `stringsAsFactors` which can be set to `FALSE`.

In most cases, it's preferable to set `stringsAsFactors = FALSE` when importing
your data, and converting as a factor only the columns that require this data
type.

Compare the output of `str(metadata)` when setting `stringsAsFactors = TRUE`
(default) and `stringsAsFactors = FALSE`:


```{r, eval=FALSE, purl=FALSE}
## Compare the difference between when the data are being read as
## `factor`, and when they are being read as `character`.
metadata <- read.csv("data/metadata.csv", sep = "\t", stringsAsFactors = TRUE)
str(metadata)
metadata <- read.csv("data/metadata.csv", sep = "\t", stringsAsFactors = FALSE)
str(metadata)
## Convert the column "Sex" into a factor
metadata$Sex <- factor(metadata$Sex)
str(metadata)
```


> ### Challenge
>
> 1. We have seen how data frames are created when using the `read.csv()`, but
>   they can also be created by hand with the `data.frame()` function.  There are
>   a few mistakes in this hand-crafted `data.frame`, can you spot and fix them?
>   Don't hesitate to experiment!
>
>     ```{r, eval=FALSE, purl=FALSE}
>     animal_data <- data.frame(animal=c("dog", "cat", "sea cucumber", "sea urchin"),
>                               feel=c("furry", "squishy", "spiny"),
>                               weight=c(45, 8 1.1, 0.8))
>     ```
>
>     ```{r, eval=FALSE, purl=TRUE, echo=FALSE}
>     ## Challenge:
>     ##  There are a few mistakes in this hand-crafted `data.frame`,
>     ##  can you spot and fix them? Don't hesitate to experiment!
>     animal_data <- data.frame(animal=c(dog, cat, sea cucumber, sea urchin),
>                               feel=c("furry", "squishy", "spiny"),
>                               weight=c(45, 8 1.1, 0.8))
>     ```
>
> 2. Can you predict the class for each of the columns in the following example?
>    Check your guesses using `str(country_climate)`:
>      * Are they what you expected?  Why? Why not?
>      * What would have been different if we had added `stringsAsFactors = FALSE` to this call?
>      * What would you need to change to ensure that each column had the accurate data type?
>
>     ```{r, eval=FALSE, purl=FALSE}
>     country_climate <- data.frame(
>            country=c("Canada", "Panama", "South Africa", "Australia"),
>            climate=c("cold", "hot", "temperate", "hot/temperate"),
>            temperature=c(10, 30, 18, "15"),
>            northern_hemisphere=c(TRUE, TRUE, FALSE, "FALSE"),
>            has_kangaroo=c(FALSE, FALSE, FALSE, 1)
>            )
>     ```
>
>    ```{r, eval=FALSE, purl=TRUE, echo=FALSE}
>    ## Challenge:
>    ##   Can you predict the class for each of the columns in the following
>    ##   example?
>    ##   Check your guesses using `str(country_climate)`:
>    ##   * Are they what you expected? Why? why not?
>    ##   * What would have been different if we had added `stringsAsFactors = FALSE`
>    ##     to this call?
>    ##   * What would you need to change to ensure that each column had the
>    ##     accurate data type?
>    country_climate <- data.frame(country=c("Canada", "Panama", "South Africa", "Australia"),
>                                   climate=c("cold", "hot", "temperate", "hot/temperate"),
>                                   temperature=c(10, 30, 18, "15"),
>                                   northern_hemisphere=c(TRUE, TRUE, FALSE, "FALSE"),
>                                   has_kangaroo=c(FALSE, FALSE, FALSE, 1))
>    ```
>
>    <!--- Answers
>
>    ```{r, eval=FALSE, echo=FALSE, purl=FALSE}
>    ## Answers
>    ## * missing quotations around the names of the animals
>    ## * missing one entry in the "feel" column (probably for one of the furry animals)
>    ## * missing one comma in the weight column
>
>    ## Answers
>    ## * `country`, `climate`, `temperature`, and `northern_hemisphere` are
>    ##    factors; `has_kangaroo` is numeric.
>    ## * using `stringsAsFactors=FALSE` would have made them character instead of
>    ##   factors
>    ## * removing the quotes in temperature, northern_hemisphere, and replacing 1
>    ##   by TRUE in the `has_kangaroo` column would probably what was originally
>    ##   intended.
>    ```
>
>    -->
>

The automatic conversion of data type is sometimes a blessing, sometimes an
annoyance. Be aware that it exists, learn the rules, and double check that data
you import in R are of the correct type within your data frame. If not, use it
to your advantage to detect mistakes that might have been introduced during data
entry (a letter in a column that should only contain numbers for instance).

<p style="text-align: right; font-size: small;">Page build on: `r format(Sys.time())`</p>
